src/ior.c

```diff
@@ -1794,53 +1811,39 @@ static IOR_offset_t *GetOffsetArrayRandom(IOR_param_t * test, int pretendRank,
-                fileSize *= test->numTasks;
-        }
+        /* count needed offsets */
+        offsets = (test->blockSize / test->transferSize) * test->segmentCount;
 
-        /* count needed offsets (pass 1) */
-        for (i = 0; i < fileSize; i += test->transferSize) {
-                if (test->filePerProc == FALSE) {
-                        // this counts which process get how many transferes in
-                        // a shared file
-                        if ((rand() % test->numTasks) == pretendRank) {
-                                offsets++;
-                        }
-                } else {
-                        offsets++;
-                }
-        }
+        /* start with same array of offsets as in sequential case */
+        offsetArray = GetOffsetArraySequential(test, pretendRank);
 
-        /* setup empty array */
-        offsetArray =
-                (IOR_offset_t *) malloc((offsets + 1) * sizeof(IOR_offset_t));
-        if (offsetArray == NULL)
-                ERR("malloc() failed");
-        offsetArray[offsets] = -1;      /* set last offset with -1 */
-
-        if (test->filePerProc) {
-                /* fill array */
-                for (i = 0; i < offsets; i++) {
-                        offsetArray[i] = i * test->transferSize;
-                }
-        } else {
-                /* fill with offsets (pass 2) */
-                srand(seed);  /* need same seed  to get same transfers as counted in the beginning*/
-                for (i = 0; i < fileSize; i += test->transferSize) {
-                        if ((rand() % test->numTasks) == pretendRank) {
-                                offsetArray[offsetCnt] = i;
-                                offsetCnt++;
-                        }
-                }
-        }
         /* reorder array */
-        for (i = 0; i < offsets; i++) {
-                value = rand() % offsets;
+        if (offsets < 2*first) first = offsets;
+        /* do first few million totally random */
+        for (i = 0; i < first; i++) {
+                value = random() % offsets;
                 tmp = offsetArray[value];
                 offsetArray[value] = offsetArray[i];
                 offsetArray[i] = tmp;
         }
+        /* do the rest in vecs chunks of vlen contiguous offsets
+           using L2 cache to reduce memory latency vs truely random */
+        for (i = first; i < offsets - 2*vecs*vlen; i += vecs*vlen) {
+                for (j = 0; j < vecs; j++) {
+                        start[j] = i + 8*(random() % (offsets - i - vlen)/8);  // align on a cache line
+                }
+                seq = 0;
+                for (k = 0; k < vlen; k++) {
+                        for (j = 0; j < vecs; j++) {
+                                value = start[j];
+                                tmp = offsetArray[value+k];
+                                offsetArray[value+k] = offsetArray[i+seq];
+                                offsetArray[i+seq] = tmp;
```

Comment:
I wonder if the three previous lines would look a bit cleaner with a helper function:
swapArrayElements(offsetArray,value+k,i+seq);
I think this would look better and be slightly easier to read.